src/radical/pilot/agent/launch_method/srun.py

```diff
@@ -207,5 +207,21 @@ def get_rank_cmd(self):
         return ret
 
 
+    # --------------------------------------------------------------------------
+    #
+    def get_env_blacklist(self):
+        '''
+        extend the base blacklist by SLURM specific environment variables
+        '''
+
+        ret = super().get_env_blacklist()
+
+        ret.extend([
+                    'SLURM_*',
```

Comment:
this should come for RM, since we can use different LMs within the slurm - we do have `rm_info` here, so maybe to add another attribute into it as `env_blacklist`? thus there are two sources for blacklists: RM and LM

Reply:
That's what we though, but I realized that this is incorrect.  The current version *does* support multiple launch methods.

Let me describe that with an example.

  - agent runs in an sbatch allocation, that sets `SLURM_NNODES=4`
  - we create a named_env called `test_env`
  - one task is started with `srun -n 2`, using that named env
  - second task is started with `mpirun -n 2`, also in that named env
  
Currently, the `prepare_env` for `test_env` will store all env, including all `SLURM_*` variables such as `SLURM_NNODES=4`.

When launching, task 1 (srun) will set `SLURM_NNODES=2`, but the named env will restore the original value (`SLURM_NNODES=4`) and the task is in trouble.

Task2 on the other hand will also have `SLURM_NNODES=4` restored, but it does not care as openmpi is not interpreting that env.  `mpirun` though will set, for example, `OMPI_RANKS=2` - that's ok as the named env does not contain any `OMPI_RANKS` setting and thus there is nothing overwritten.

With the current code in this PR, the following happens: on `prepare_env`, we prepare the env [**individually for each launch method**](https://github.com/radical-cybertools/radical.pilot/pull/3203/files#diff-24efcf4c13da0181a5726fc2b70567d3040a49decb3925b3ae322a78c74d697bL204).  So when first using `srun` to launch task 1, that code will get the blacklist from `srun` which will tell `don't store "SLURM_NNODES" because srun will set it itself, leading to a conflict`.  The named env will not contain that env, and the task get's the correct env from `srun`.

When starting task 2 with mpirun, the same code will ask the mpirun LM for the blacklist.  That will say `don't store "OMPI_RANKS" because conflict!` - which is fine: the agent env did not contain `OMPI_RANKS` in the first place, so all happens as before: `SLURM_NNODES` is kept in the env but does not pose any problems.

So essentially, what the blacklist says is: "this launch methods needs to set `THIS_LM_FOO`, so don't store that in any named env to avoid conflicts".  It does not really matter if the env contains that env setting or not - and that is why we don't need to ask the RM about it either, it would not make any difference.

Reply:
sorry, I don't disagree with the concept of blacklisting, just propose to keep corresponding lists where they belong, env variables that are set by a batch system should be within RMInfo, so for Slurm RM it will be `{... 'env_blacklist': ['SLURM_*']}`, while for PBSPro RM will `{... 'env_blacklist': ['PBS_*']}`

and then we use it within LM, and will have the following in base class `LaunchMethod.get_env_blacklist`

```python
        ret = [
                 # RP internal variables
                 'RP_*',

                 # resource details
                 'ROCR_VISIBLE_DEVICES',
                 'CUDA_VISIBLE_DEVICES',
         ]
        ret.extend(self._rm_info['env_blacklist'])
        return ret
```

Other LMs will inherit it and expand it with their related env variables (for now it is just rank-related)

Reply:
Sorry for being unclear, but that was my point: they *belong* into the LM!  The *SRUN LM* will set `SLURM_NNODES` and will thus need to have that blacklisted from overwriting, the `MPIRUN LM` will set `OMPI_RANKS` and thus needs to have that blacklisted instead.  It does not matter what the RM thinks needs blacklisting - the *LM* needs to decide what it needs to trickle through to the task.

In you example, it does not make  sense to unset `SLURM*` when using mpirun as LM - why would we do that?  

Remember, the goal is to recreate what a user would see when running a batch script.  in that case (`sbatch` script and then `mpirun`), the `SLURM_*`  would not be unset, and `mpirun` would work all right.  With `sbatch + srun`, the `SLURM_*` setting from srun would not be unset, but instead the `SLURM_*` settings from `srun` would overwrite them.  So that is what we recreate: let the `srun LM` report what vars it will overwrite (and thus need blacklisting from env caching).